memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -208,44 +210,43 @@ func (r *ReconcileMemcached) deploymentForMemcached(m *cachev1alpha1.Memcached)
 				},
 				Spec: corev1.PodSpec{
 					Containers: []corev1.Container{{
-						Image:   "memcached:1.4.36-alpine",
-						Name:    "memcached",
-						Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
-						Ports: []corev1.ContainerPort{{
-							ContainerPort: 11211,
-							Name:          "memcached",
-						}},
+						Name:    name,
+						Image:   image,
+						Ports:   ports,
+						Command: command,
 					}},
 				},
 			},
-		},
+		}
+
+		return
 	}
-	// Set Memcached instance as the owner of the Deployment.
-	controllerutil.SetControllerReference(m, dep, r.scheme)
-	return dep
+
+	// Ensure all fields in the Deployment have their expected values. If they're not already
+	// set to these values, we reset them here and then when control is given back to the
+	// `controllerutil.CreateOrUpdate` function, it will send that update to the api server.
+	deployment.Spec.Replicas = &replicas
+	deployment.Spec.Template.ObjectMeta.Labels = ls
+	deployment.Spec.Template.Spec.Containers[0].Name = name
```

Comment:
Could not this point face an expection deployment.Spec.Template.Spec.Containers[0]?

Reply:
Yeah, that's what I was hinting at in my comment after I made this change yesterday (saying that it might be a bit naive to assume it will be there). I'll update this. :+1:

Reply:
I think the correct way to approach this to handle both the create and update cases is to do something like this:

```go
if deployment.Spec == nil {
    deployment.Spec = &appsv1.DeploymentSpec{}
}
if deployment.Spec.Selector == nil {
    deployment.Spec.Selector = &metav1.LabelSelector{}
}
if deployment.Spec.Template == nil {
    deployment.Spec.Template = &corev1.PodTemplateSpec{}
}
if deployment.Spec.Template.ObjectMeta == nil {
    deployment.Spec.Template.ObjectMeta = &metav1.ObjectMeta{}
}
if deployment.Spec.Template.Spec == nil {
    deployment.Spec.Template.Spec = &corev1.PodSpec{}
}
if len(deployment.Spec.Template.Spec.Containers) != 1 {
    deployment.Spec.Template.Spec.Containers = make([]corev1.Container, 1)
}
deployment.Spec.Replicas = &replicas
deployment.Spec.Selector.MatchLabels = ls
deployment.Spec.Template.ObjectMeta.Labels = ls
deployment.Spec.Template.Spec.Containers[0].Name = name
deployment.Spec.Template.Spec.Containers[0].Image = image
deployment.Spec.Template.Spec.Containers[0].Ports = ports
deployment.Spec.Template.Spec.Containers[0].Command = command
```

Reply:
@joelanford for most of the nil checks in your example, the compiler complains with e.g.

```
cannot convert nil to type "k8s.io/api/apps/v1".DeploymentSpec
```

so I think those ones are then not needed, right (zero value is the empty struct already)?

The `deployment.Spec.Selector` seems to be a very odd case indeed:
- it's a pointer rather than a value, so it doesn't error at compile time
- It _must_ contain a non-empty `matchLabels` field
- It cannot be deleted accidentally (if I try to remove it from the deployment manually, it gets accepted, but it seems like that update just gets silently dropped (watchers don't see any change)
- While the labels can't all be removed manually, all-but-one can, and _that_ update will be accepted. However, back in the operator, even trying to ensure the labels are individually there, e.g.

```go
for k, v := range ls {
	deployment.Spec.Selector.MatchLabels[k] = v
}
```

results in an error like this:

<details>
{"level":"error","ts":1571261523.3205278,"logger":"controller_memcached","msg":"Failed to reconcile Deployment","Request.Namespace":"memcached","Request.Name":"example-memcached","Deployment.Namespace":"memcached","Deployment.Name":"example-memcached","error":"Deployment.apps \"example-memcached\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app\":\"memcached\", \"memcached_cr\":\"example-memcached\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/controller/memcached.(*ReconcileMemcached).Reconcile\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/controller/memcached/memcached_controller.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/gryan/go/src/github.com/operator-framework/operator-sdk-samples/memcached-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
</details>

So I (`system:admin` in OpenShift v3.11 at least) can manually mess up the `deployment.Spec.Selector.MatchLabels` to a certain degree with `kubectl edit deploy example-memcached`, but the operator can't fix that. At least I can't think of how the operator might be able to do it, without the nuclear option of deleting the Deployment altogether, and creating a new one.

I also notice that the godocs for controller-runtime have a code example for `CreateOrUpdate`, and that's working with a Deployment, and handles the `Spec.Selector` differently than everything else: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate